modules/light-clients/08-wasm/types/misbehaviour_handle_test.go

```diff
@@ -469,372 +483,131 @@ package types_test
-				suite.Require().NoError(err)
-				clientMsg = &types.ClientMessage{
+				clientMessage = &types.ClientMessage{
 					Data: data,
 				}
 
-				err = clientState.VerifyClientMessage(suite.ctx, suite.chainA.Codec, suite.store, clientMsg)
-				suite.Require().NoError(err)
+				suite.mockVM.RegisterQueryCallback(types.CheckForMisbehaviourMsg{}, func(codeID wasmvm.Checksum, env wasmvmtypes.Env, queryMsg []byte, store wasmvm.KVStore, goapi wasmvm.GoAPI, querier wasmvm.Querier, gasMeter wasmvm.GasMeter, gasLimit uint64, deserCost wasmvmtypes.UFraction) ([]byte, uint64, error) {
+					resp := fmt.Sprintf(`{"found_misbehaviour":%t}`, true)
+					return []byte(resp), types.DefaultGasUsed, nil
+				})
+
+				foundMisbehaviour = clientState.CheckForMisbehaviour(
+					suite.chainA.GetContext(),
+					suite.chainA.App.AppCodec(),
+					clientStore,
+					clientMessage,
+				)
 			},
 			true,
+			true,
 		},
 		{
-			"invalid wasm misbehaviour",
-			func() {
-				clientMsg = &solomachine.Misbehaviour{}
+			"query failed, vm panic", func() {
+				data, err := base64.StdEncoding.DecodeString(suite.testData["header"])
+				suite.Require().NoError(err)
+				clientMessage = &types.ClientMessage{
+					Data: data,
+				}
+
+				suite.mockVM.RegisterQueryCallback(types.CheckForMisbehaviourMsg{}, func(codeID wasmvm.Checksum, env wasmvmtypes.Env, queryMsg []byte, store wasmvm.KVStore, goapi wasmvm.GoAPI, querier wasmvm.Querier, gasMeter wasmvm.GasMeter, gasLimit uint64, deserCost wasmvmtypes.UFraction) ([]byte, uint64, error) {
+					resp := "cannot be unmarshalled"
+					return []byte(resp), types.DefaultGasUsed, nil
+				})
+
+				foundMisbehaviour = clientState.CheckForMisbehaviour(
+					suite.chainA.GetContext(),
+					suite.chainA.App.AppCodec(),
+					clientStore,
+					clientMessage,
+				)
 			},
 			false,
+			false,
 		},
 	}
 
 	for _, tc := range testCases {
 		suite.Run(tc.name, func() {
-			suite.SetupWasmGrandpaWithChannel()
-			clientState, ok = suite.chainA.App.GetIBCKeeper().ClientKeeper.GetClientState(suite.ctx, grandpaClientID)
-			suite.Require().True(ok)
+			// reset suite to create fresh application state
+			suite.SetupWasmWithMockVM()
 
-			tc.malleate()
+			endpoint := wasmtesting.NewWasmEndpoint(suite.chainA)
+			err := endpoint.CreateClient()
+			suite.Require().NoError(err)
 
-			foundMisbehaviour := clientState.CheckForMisbehaviour(suite.ctx, suite.chainA.Codec, suite.store, clientMsg)
+			clientStore = suite.chainA.App.GetIBCKeeper().ClientKeeper.ClientStore(suite.chainA.GetContext(), endpoint.ClientID)
+			clientState = endpoint.GetClientState()
 
 			if tc.expPass {
-				suite.Require().True(foundMisbehaviour)
+				suite.Require().NotPanics(tc.checkMisbehaviour, "unexpected panic")
+				suite.Require().Equal(tc.foundMisbehaviour, foundMisbehaviour)
 			} else {
-				suite.Require().False(foundMisbehaviour)
+				suite.Require().Panicsf(tc.checkMisbehaviour, "failed to unmarshal result of wasm query")
```

Comment:
i think the reason we panic is because the return value of this function is `bool`, so we can't really throw an error (because we are mocking the tendermint vm/functions, i guess we wouldnt change this function signature)